### ISSUE TITLE
[0.0.78] Bot buys but doesn't sell 

### ISSUE THREAD

**Issue by ilbuonmarcio** [chrisleekr/binance-trading-bot#298]
## Expected Behaviours

The bot sells and buys automatically.

# Bugs found

I expect also the second coin from the left (SXP) to show the Trigger button, and also the ability to sell, but it jumps up and down the trigger and doesn't trigger once.
It just happened also for FTT (Which had +21% which I managed to trigger manually via Market sell, and XLM on +9% via Limit sell).

It seems to happen just for some coins and not others.
FTT was purchased today automatically via the bot at 5AM UTC time, and it hasn't triggered sell signal automatically.

Is there something I need to know?
Sell is at the following settings: 1.0525, 0.989, 0.985, 1

## Screenshots

![immagine](https://user-images.githubusercontent.com/21985554/132246420-8367af7e-1fb6-4c93-8ff2-10d3492aca94.png)

## Additional context

Bot v.0.0.77, on cloud VPS, 2 core 2gb ram x86, "standard configuration"

**Comment by ilbuonmarcio** [comment]
I just noticed that I don't have Trigger green and red button for almost 70% of the coins, what could be the problem?

**Comment by ilbuonmarcio** [comment]
After some trial and error, and after upgrading to v0.0.78, still nothing changed.

I had 35 USDT free for almost 40 minutes, and thought they will be used in some buy trades but they were never used in this timeframe.

One thing I noticed is that if I select a coin which doesn't show up Trigger red/green button, if I go into its specific settings and click under ACTIONS -> Reset Grid Trade and click Archive and Delete, the bot instantly releases and shows up the Red/Green button! And in case of a possible coin buy, it instantly triggers a buy signal and opens an order... as expected...

@chrisleekr is that a possible bug? What can be the cause of that?

**Comment by chrisleekr** [comment]
@ilbuonmarcio 

1. What errors do you get from Slack?
2. Can you check the log?
3. Is this a similar issue - https://github.com/chrisleekr/binance-trading-bot/issues/284?

**Comment by ilbuonmarcio** [comment]
No errors on Slack whatsoever, just normal logs about login, sell and buy orders.

The Issue you mentioned I had previously in the past but it's not the case now

**Comment by ilbuonmarcio** [comment]
What should I do to check the logs properly?

docker-compose logs -f --tail=100

I usually use a command like this, but there is so much data it becomes unreasonable sadly

**Comment by chrisleekr** [comment]
Oh wait, I checked the screenshot again. 

SXPUSDT Sell Grid Trade is already executed, but you still have enough USDT, hence the bot didn't remove the last buy price.

![image](https://user-images.githubusercontent.com/5715919/132260869-065bdf46-a22c-499d-a1f6-96d99b1cb066.png)

Can you try to access mongodb and get the grid trade order?

```bash
$ docker exec -it binance-mongo /bin/bash 
> use binance-bot
switched to db binance-bot
> db['trailing-trade-grid-trade'].find({"key":"SXPUSDT"});
```

**Comment by ilbuonmarcio** [comment]
I only have one "level" of grid trade configuration both for sell and buy, is that any relevant?

I'm checking mongo as per your suggestions right now

**Comment by chrisleekr** [comment]
> docker-compose logs -f --tail=100
>I usually use a command like this, but there is so much data it becomes unreasonable sadly

For getting a specific coin log, try this way.

```
$ docker logs binance-bot --tail=100 -f | bunyan -c 'this.symbol === "SXPUSDT"' > binance.log
```

Then open binance.log


> I only have one "level" of grid trade configuration both for sell and buy, is that any relevant?

No, it shouldn't be the issue. 

You configured sell grid trade as 100%, it should've sold all of your coins but left 4.3 SXP.
I assume you didn't manually purchase SXP outside of the bot. So I need to figure out why it didn't sell everything.


I want to see

**Comment by ilbuonmarcio** [comment]
No rows response shows up as per your suggestions about mongo, maybe because I already forced the "fix" I found and applied it for all coins already...

I never bought crypto outside of the bot whatsoever, except for USDT for feeding the bot.

**Comment by ilbuonmarcio** [comment]
My buy config is buying 16 USDT max, so it's not "leaving" any coins behind, it's just all that it has ever bought in the trade buy/sell session

**Comment by chrisleekr** [comment]
> No rows response shows up as per your suggestions about mongo, maybe because I already forced the "fix" I found and applied it for all coins already...

Oh did you already reset the grid trade for SXPUSDT?

Is this issue happening now to other coins? 

What is the current issue?

**Comment by ilbuonmarcio** [comment]
The bot should not reset the last buy price because the price didn't fall under 10 USDT, it just went up from 16 to the value you see on the latest screenshot.

This problem occurs even with coins on buy (the Trigger button doesn't show up and the bot was blocked at the "will not buy this coin even with USDT available" state) but which never triggered a buy signal due to this "block" I was experiencing with my bot, not just sell time.

To recap, I already reset grid trade for all coins which didn't show up Trigger manual sell/buy button, and they started selling and buying correctly after that.

Due to having more than two bots currently working on the exact identical configuration, and only experiencing this problem on one singular instance, that seems reasonable to debug the situation

**Comment by ilbuonmarcio** [comment]
To be fair, the only difference between the other instances and this instance is that this instance has 3 to 5 more crypto USDT pairs to watch, but that doesn't seem to be the issue here I think

**Comment by ilbuonmarcio** [comment]
I'm using the bot since June 21th, and all other bots working good since then (other than the tickSize issue you referenced earlier), so I'm not convinced I made a configuration mistake between this and the other ones, but let's be fair, it could possibly be me but I'm not sure about what went wrong :)

**Comment by chrisleekr** [comment]
> The bot should not reset the last buy price because the price didn't fall under 10 USDT, it just went up from 16 to the value you see on the latest screenshot.
> This problem occurs even with coins on buy (the Trigger button doesn't show up and the bot was blocked at the "will not buy this coin even with USDT available" state) but which never triggered a buy signal due to this "block" I was experiencing with my bot, not just sell time.

Yeah, something happened in there unless I can see it now, I won't be able to tell what went wrong. 

> To recap, I already reset grid trade for all coins which didn't show up Trigger manual sell/buy button, and they started selling and buying correctly after that.

Ok, that is good. So it is working now. 

> Due to having more than two bots currently working on the exact identical configuration, and only experiencing this problem on one singular instance, that seems reasonable to debug the situation
> To be fair, the only difference between the other instances and this instance is that this instance has 3 to 5 more crypto USDT pairs to watch, but that doesn't seem to be the issue here I think

Wait, are you running two bots on the same server?
Are they sharing MongoDB/Redis? Are they having separate MongoDB/Redis instances or separate MongoDB DB /Redis logical DB. 

> I'm using the bot since June 21th, and all other bots working good since then (other than the tickSize issue you referenced earlier), so I'm not convinced I made a configuration mistake between this and the other ones, but let's be fair, it could possibly be me but I'm not sure about what went wrong :)

Next time if it happens, 
1. Get the logs as per the above instructions
2. Get Slack messages

If you have a log, you would figure out what is going on. 
I understand the log is extensive. But it definitely helps to figure out what is going on.

**Comment by ilbuonmarcio** [comment]
I am running multiple bots on multiple virtual machines on a cloud provider, one bot instance and IP address per machine, they are completely isolated from each other.

In the upcoming days I'm gonna monitor the other bots (which are currently on 0.0.76, I'm gonna migrate them to 0.0.78), I'll let you know if I face problems like this, and if so I'll not do the "fix" I found and gently get in touch with you with the appropriate loggings you asked for.

The only changes I did for your bot, in the docker-compose.yml file (I usually run the master version) I changed the node, redis and mongo bindings from 8080:80 to 127.0.0.1:8080:80 for example, in order to isolate the services from the outside, and install nginx as a http proxy to access the bot via that; also, I changed the main app log size from 1G to 50M. Other than that, it's the same exact config as master branch is showing right now.

**Comment by chrisleekr** [comment]
> I am running multiple bots on multiple virtual machines on a cloud provider, one bot instance and IP address per machine, they are completely isolated from each other.

Oh ok, that is ok. 

> In the upcoming days I'm gonna monitor the other bots (which are currently on 0.0.76, I'm gonna migrate them to 0.0.78), I'll let you know if I face problems like this, and if so I'll not do the "fix" I found and gently get in touch with you with the appropriate loggings you asked for.

Yap, yap. Let me know.

> The only changes I did for your bot, in the docker-compose.yml file (I usually run the master version) I changed the node, redis and mongo bindings from 8080:80 to 127.0.0.1:8080:80 for example, in order to isolate the services from the outside, and install nginx as a http proxy to access the bot via that; 

Wait, why are you using `docker-compose.yml`, not `docker-compose.server.yml`?  Any specific reason? Unless you do the development, you don't need to use `docker-compose.yml`.

> also, I changed the main app log size from 1G to 50M. Other than that, it's the same exact config as master branch is showing right now.

That should be ok.

**Comment by ilbuonmarcio** [comment]
I always used the standard `docker-compose build && docker-compose up -d --build` command to run the bot from the beginning of times, so it used `docker-compose.yml`, no apparent reason to be honest.

Is the other a more production-ready solution? If so, can this be the problem?

**Comment by chrisleekr** [comment]
> I always used the standard docker-compose build && docker-compose up -d --build command to run the bot from the beginning of times, so it used docker-compose.yml, no apparent reason to be honest.
> Is the other a more production-ready solution? If so, can this be the problem? I mean, using the development version for such long periods

It shouldn't the issue. The thing is, if you are using `docker-compose.server.yml`, you won't need to build at all, which takes some time. Just pulling a new image should be enough. 

Since v0.0.76, I have made some improvements in the frontend - https://github.com/chrisleekr/binance-trading-bot/blob/master/CHANGELOG.md#0076---2021-08-17
So using the production image unless you are doing the development is making more sense to have a benefit of the performance. 
But you said you are using Cloud, so you may not feel much difference since that should be always fast. 

I think what you have experienced is, it may be caused by some migrations issue. Most likely my mistake lol. But again, I am not sure why it happened without seeing your log. 

Anyway, you seem to know what you are doing. 
Let's close the issue for now and open a new one if you have the same issue.

**Comment by ilbuonmarcio** [comment]
Yeah, sure! 
Thanks for the support, always appreciated.

I'm planning on making future contributions to the project, just to let you know; it's not my main stack of choice, and it has something new to learn in it, but if contributions are accepted I'm going to be here for a while :)

**Comment by chrisleekr** [comment]
@ilbuonmarcio Welcome any contributions :)

**Comment by ilbuonmarcio** [comment]
I just updated all the instances, no problems arised whatsoever, it just had been a coincidence then, good! 💯
